File: julia/euler4.jl
#=

  Euler #4 in Julia.

  Problem 4
  """
  A palindromic number reads the same both ways. The largest palindrome made
  from the product of two 2-digit numbers is 9009 = 91 × 99.

  Find the largest palindrome made from the product of two 3-digit numbers.
  """


  This Julia program was created by Hakan Kjellerstrand, hakank@gmail.com
  See also my Julia page: http://www.hakank.org/julia/

=#
include("Euler.jl")

# 0.00958293s
function euler4a()
    max = 0
    from = 100
    to = 999
    for i in from:to, j in i:to
        ij = i*j
        if ij > max && palindromic_number(ij)
            max = ij
        end
    end
    return max;
end

# 0.313482985s
function euler4b()
    from = 100
    to = 999
    a = []

    # Note: It seems that one cannot assign a variable (ij=i*j)
    # in a function pipe
    (from:to).|> i->
        (i:to).|> j->
            # ij = i * j; # this don't work! (Why?)
            if palindromic_number(i*j)
                push!(a,i*j)
            end
    return maximum(a)
end

# 0.085608391s
function euler4c()
    from = 100
    to = 999
    # vcat(a...) flattens the array
    a = (from:to).|>i->(i:to).|>j->i*j
    return maximum(filter(ij->palindromic_number(ij),vcat(a...)))
end

run_euler(euler4a)
# run_euler(euler4b)
# run_euler(euler4c)

File: haskell/euler4.hs
{- 
  
  Euler #4 in Haskell

  Problem 4
  """
  A palindromic number reads the same both ways. The largest palindrome made 
  from the product of two 2-digit numbers is 9009 = 91 × 99.

  Find the largest palindrome made from the product of two 3-digit numbers.
  """


  This program was created by Hakan Kjellerstrand (hakank@gmail.com)
  Also see my Haskell page: http://hakank.org/haskell/
   
-}

import Data.List
import HakankUtils

-- palin xs = xs == reverse xs

euler4a = maximum [x*y | x <- [100..999], y <- [x..999], palin (show (x*y))]
euler4b = maximum [t | x <- [100..999], y <- [x..999], t <- [x*y], palin (show t)]

-- Using Applicative Function for cross product: Slower
euler4c = maximum $ filter (\x -> palin $ show x) ( pure (*) <*> [100..999] <*> [100..999])

main = do
         print euler4a -- (0.29 secs, 317,132,720 bytes)
         -- print euler4b -- (0.34 secs, 359,220,224 bytes)
         -- print euler4c -- (0.44 secs, 581,379,552 bytes)

         
